WalletWasabi.Tests/UnitTests/WabiSabi/Client/CredentialDependencyTests.cs

```diff
@@ -25,99 +25,45 @@ public async void AsyncDependencyGraphTraversalAsync()
+			var taskCompletionSources = allEdges.ToImmutableDictionary(e => e, _ => new TaskCompletionSource<ulong>());
 
-			// Also keep track of the in-flight requests
-			var inFlightRequests = new List<(Task<ImmutableSortedDictionary<CredentialType, IEnumerable<ulong>>> Task, ImmutableSortedDictionary<CredentialType, IEnumerable<CredentialDependency>> Dependencies)>();
-
-			// And all sent requests, for testing purposes.
-			var sent = new HashSet<RequestNode>();
+			// Track all "sent" requests for testing purposes
+			var sent = new ConcurrentDictionary<RequestNode, bool>();
 
 			var rng = new Random();
 
 			// Simulate sending a request. Instead of actual Credential objects,
 			// credentials are just represented as ulongs.
-			async Task<ImmutableSortedDictionary<CredentialType, IEnumerable<ulong>>> SimulateRequest(
-				RequestNode node,
-				ImmutableSortedDictionary<CredentialType, IEnumerable<ulong>> presented,
-				ImmutableSortedDictionary<CredentialType, IEnumerable<ulong>> requested)
-			{
-				foreach (var credentialType in presented.Keys)
-				{
-					Assert.Equal(g.InEdges(node, credentialType).Select(e => e.Value).OrderBy(x => x), presented[credentialType].OrderBy(x => x));
-					Assert.Equal(g.OutEdges(node, credentialType).Select(e => e.Value).OrderBy(x => x), requested[credentialType].OrderBy(x => x));
-				}
-
-				Assert.DoesNotContain(node, sent);
-				sent.Add(node);
-
-				await Task.Delay(1 + rng.Next(10));
-
-				return requested;
-			}
-
-			var ct = new CancellationTokenSource(new TimeSpan(0, 2, 0));
-
-			for (var remainingSteps = 2 * pendingCredentialsToPresent.Count; remainingSteps > 0 && pendingCredentialsToPresent.Count + inFlightRequests.Count > 0; remainingSteps--)
+			async Task SimulateRequest(RequestNode node)
 			{
-				// Clear unblocked but waiting requests. Not very efficient
-				// (quadratic complexity), but good enough for demonstration
-				// purposes.
-				foreach (var node in UnblockedRequests())
-				{
-					// FIXME this is a little ugly, how should it look in the real code? seems like we're missing an abstraction
-					var edgesByType = DependencyGraph.CredentialTypes.ToImmutableSortedDictionary(t => t, t => g.OutEdges(node, t));
-					var credentialsToRequest = edgesByType.ToImmutableSortedDictionary(kvp => kvp.Key, kvp => kvp.Value.Select(e => e.Value));
-					var credentialsToPresent = pendingCredentialsToPresent[node].ToImmutableSortedDictionary(kvp => kvp.Key, kvp => kvp.Value.AsEnumerable());
+				var inEdges = Enumerable.Concat(g.InEdges(node, CredentialType.Amount), g.InEdges(node, CredentialType.Vsize));
 
-					Assert.NotEmpty(edgesByType);
+				var credentialTasks = inEdges.Select(edge =>taskCompletionSources[edge].Task);
+				await Task.WhenAll(credentialTasks);
 
-					var task = SimulateRequest(node, credentialsToPresent, credentialsToRequest);
+				var credentialsToPresent = credentialTasks.Select(t => t.Result);
```

Comment:
Here there are two alternatives:

You can use the results returned by the `Task.WhenAll`:

```suggestion
				var credentialsToPresent = await Task.WhenAll(credentialTasks);
```

Or `await` the task because in case it is completed (it is) them you will get the result but, in case it is failed you will get an exception thrown.
```suggestion
				var credentialsToPresent = credentialTasks.Select(t => wait t);
```

Reply:
I can be merged with or without this.

Reply:
which do you prefer? an exception should be thrown if anything fails, i was kind of mindlessly copying existing code. i think the second is missing an `async` in the lambda and also has a typo, and thanks to github the first leaves an extra `WhenAll` call (although harmless?) because the empty line makes it impossible to suggest combining those lines, so i will apply manually

Reply:
i think the first one is better. In case of failure it should throw and aggregated exception it think.